daml-assistant/integration-tests/src/DA/Daml/Assistant/IntegrationTests.hs

```diff
@@ -545,27 +545,62 @@ codegenTests codegenDir = testGroup "daml codegen" (
           callCommandSilent "daml new create-daml-app create-daml-app"
         let cdaDir = tmpDir </> "create-daml-app"
+
+        -- First test the base application (without the user-added feature).
         withCurrentDirectory cdaDir $ do
+          step "Build DAML model for base application"
           callCommandSilent "daml build"
+          step "Set up TypeScript libraries and Yarn workspaces for codegen"
           setupYarnEnv tmpDir (Workspaces ["create-daml-app/daml.js"]) [DamlTypes]
+          step "Run JavaScript codegen"
           callCommandSilent "daml codegen js -o daml.js .daml/dist/create-daml-app-0.1.0.dar"
         assertFileDoesNotExist (cdaDir </> "ui" </> "build" </> "index.html")
         withCurrentDirectory (cdaDir </> "ui") $ do
           -- NOTE(MH): We set up the yarn env again to avoid having all the
           -- dependencies of the UI already in scope when `daml2js` runs
           -- `yarn install`. Some of the UI dependencies are a bit flaky to
           -- install and might need some retries.
+          step "Set up libraries and workspaces again for UI build"
           setupYarnEnv tmpDir (Workspaces ["create-daml-app/ui"]) allTsLibraries
+          step "Install dependencies for UI"
           retry 3 (callCommandSilent "yarn install")
+          step "Run linter"
           callCommandSilent "yarn lint --max-warnings 0"
+          step "Build the application UI"
           callCommandSilent "yarn build"
         assertFileExists (cdaDir </> "ui" </> "build" </> "index.html")
 
+        -- Now test that the messaging feature works by applying the necessary
+        -- changes and testing in the same way as above.
+        step "Patch the application code with messaging feature"
+        messagingPatch <- locateRunfiles (mainWorkspace </> "templates" </> "messaging.patch")
+        patchTool <- locateRunfiles "patch_dev_env/bin/patch"
+        withCurrentDirectory cdaDir $ do
+          callCommandSilent $ unwords [patchTool, "-s", "-p2", "<", messagingPatch]
+          forM_ ["MessageEdit", "MessageList"] $ \messageComponent ->
+            assertFileExists ("ui" </> "src" </> "components" </> messageComponent <.> "tsx")
+          step "Build the new DAML model"
+          callCommandSilent "daml build"
+          step "Set up TypeScript libraries and Yarn workspaces for codegen again"
+          setupYarnEnv tmpDir (Workspaces ["create-daml-app/daml.js"]) [DamlTypes]
+          step "Run JavaScript codegen for new DAML model"
+          callCommandSilent "daml codegen js -o daml.js .daml/dist/create-daml-app-0.1.0.dar"
+        withCurrentDirectory (cdaDir </> "ui") $ do
+          step "Set up libraries and workspaces again for UI build"
+          setupYarnEnv tmpDir (Workspaces ["create-daml-app/ui"]) allTsLibraries
+          step "Install UI dependencies again, forcing rebuild of generated code"
+          retry 3 (callCommandSilent "yarn install --force --frozen-lockfile")
```

Comment:
Do we still expect this to be flaky here? I would have thought that the flaky part is getting dependencies from the network. This shouldn’t happen here. Unless I’m missing something here, we should remove the `retry 3`.

Reply:
Good point, will do